gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +440,66 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
     return null;
   }
 
+  @VisibleForTesting
+  ChannelCredentials createMtlsToS2AChannelCredentials(
+      InputStream trustBundle, InputStream privateKey, InputStream certChain) throws IOException {
+    if (trustBundle == null || privateKey == null || certChain == null) {
+      return null;
+    }
+    return TlsChannelCredentials.newBuilder()
+        .keyManager(privateKey, certChain)
+        .trustManager(trustBundle)
+        .build();
+  }
+
+  ChannelCredentials createS2ASecuredChannelCredentials() {
+    S2A s2aUtils = S2A.newBuilder().build();
+    String plaintextAddress = s2aUtils.getPlaintextS2AAddress();
+    String mtlsAddress = s2aUtils.getMtlsS2AAddress();
+    if (!mtlsAddress.isEmpty()) {
+      // Currently, MTLS to MDS is only available on GCE. See:
+      // https://cloud.google.com/compute/docs/metadata/overview#https-mds
+      // Try to load MTLS-MDS creds.
+      InputStream trustBundle = null;
+      InputStream privateKey = null;
+      InputStream certChain = null;
+      try {
+        trustBundle = new FileInputStream(MTLS_MDS_ROOT);
+        privateKey = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+        certChain = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+      } catch (FileNotFoundException ignore) {
+        // Fallback to plaintext-to-S2A connection.
+        LOG.log(
+            Level.INFO,
+            "Cannot establish an mTLS connection to S2A due to error loading MTLS to MDS credentials, falling back to plaintext connection to S2A: "
+                + ignore.getMessage());
+      }
+      ChannelCredentials mtlsToS2AChannelCredentials = null;
+      try {
+        // Try to connect to S2A using mTLS.
+        mtlsToS2AChannelCredentials =
+            createMtlsToS2AChannelCredentials(trustBundle, privateKey, certChain);
+      } catch (IOException ignore) {
+        // Fallback to plaintext-to-S2A connection.
+        LOG.log(
+            Level.INFO,
+            "Cannot establish an mTLS connection to S2A due to error creating MTLS to MDS TlsChannelCredentials credentials, falling back to plaintext connection to S2A: "
+                + ignore.getMessage());
+      }
+      if (mtlsToS2AChannelCredentials != null) {
+        return S2AChannelCredentials.newBuilder(mtlsAddress, mtlsToS2AChannelCredentials).build();
+      }
+    }
+
+    if (!plaintextAddress.isEmpty()) {
+      // Fallback to plaintext connection to S2A.
+      return S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
+          .build();
+    }
```

Comment:
Is this supposed to be the last check? Shouldn't this be one of the earliest checks? I assumed so since we are retrieving the S2AConfig in the first step.

Reply:
`plaintextAddress` is a fallback, so we only really want to use if we fail to use `mtlsAddress` to create mtls to S2A channel credentials, which is why the check is at the very end. I did some refactoring in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/30a37c221432cbf312ccdc8728dad5da23cf316e which may help the readability.